diff --git a/pc/channelmanager.cc b/pc/channelmanager.cc
index eda5a2d..84bc1de 100644
--- a/pc/channelmanager.cc
+++ b/pc/channelmanager.cc
@@ -94,12 +94,14 @@ void ChannelManager::GetSupportedVideoCodecs(
 
   std::vector<VideoCodec> video_codecs = media_engine_->video().codecs();
   for (const auto& video_codec : video_codecs) {
+    if( video_codec.name == kH264CodecName ){
     if (!enable_rtx_ &&
         absl::EqualsIgnoreCase(kRtxCodecName, video_codec.name)) {
       continue;
     }
     codecs->push_back(video_codec);
   }
+  }
 }
 
 void ChannelManager::GetSupportedVideoRtpHeaderExtensions(
